ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java

```diff
@@ -145,8 +145,7 @@ public void onMatch(RelOptRuleCall call) {
         int fieldCount = joinInput.getRowType().getFieldCount();
         final ImmutableBitSet fieldSet =
             ImmutableBitSet.range(offset, offset + fieldCount);
-        final ImmutableBitSet belowAggregateKeyNotShifted =
-            belowAggregateColumns.intersect(fieldSet);
```

Comment:
@jcamachor 
I think this intersect is not needed; it cause some trouble in case
* grouped by say 1,3,4
* columns 1,2,3 is coming from left side of join; column 2 is the join key
* the intersect will remove column 2 ; which will induce a shift in the mapping - and will rotate in the join key into a projkect later on

I think in calcite the following is true:
for a join betwen inputs: 
* T1(a,b,c)
* T2(d,e,f)
the output is always a,b,c,d,e,f - independently the join condition
so I see no rational need for this `intersect`...it was here for a few years now...and the class "borned" with this line in it...

Reply:
`belowAggregateColumns` includes the aggregate fields for all inputs together with the join keys (L120). Intersection is supposed to gather only those applicable to current input in the loop, it should not remove column 2 unless I am missing anything...

Reply:
right; the column which was causing the trouble was the preceeding join's joinkey.

the issue was caused by
* that between the previous / and current join there were no projects - so all the join keys of the previous join were present in the input
* meanwhile the aggregate had references to `0,2,...` columns - which were unique; so the logic assumed that the `joinInput` could be used as is

however..because column 1 was present in the input; but not in the output this have caused that the actual join was not in sync with the mapping being created

I think the patch may make more sense than the above reasoning :)

Reply:
Yes, this makes sense now!